oidc-exchange.py

```diff
@@ -0,0 +1,161 @@
+
+This generally indicates a workflow configuration error, such as insufficient
+permissions. Make sure that your workflow has `id-token: write` configured
+at either the workflow or job level, e.g.:
+
+```yaml
+permissions:
+  id-token: write
+```
+"""
+
+# Rendered if the package index refuses the given OIDC token.
+_SERVER_REFUSED_TOKEN_EXCHANGE_MESSAGE = """
+Token request failed: the server refused the request for the following reasons:
+
+{reasons}
+"""
+
+# Rendered if the package index's token response isn't valid JSON.
+_SERVER_TOKEN_RESPONSE_MALFORMED_JSON = """
+Token request failed: the index produced an unexpected
+{status_code} response.
+
+This strongly suggests a server configuration or downtime issue; wait
+a few minutes and try again.
+"""
+
+# Rendered if the package index's token response isn't a valid API token payload.
+_SERVER_TOKEN_RESPONSE_MALFORMED_MESSAGE = """
+Token response error: the index gave us an invalid response.
+
+This strongly suggests a server configuration or downtime issue; wait
+a few minutes and try again.
+"""
+
+
+def die(msg: str) -> NoReturn:
+    with _GITHUB_STEP_SUMMARY.open("a", encoding="utf-8") as io:
+        print(msg, file=io)
+
+    # NOTE: `msg` is Markdown formatted, so we emit only the header line to
+    # avoid clogging the console log with a full Markdown formatted document.
+    header = msg.splitlines()[0]
+    print(f"::error::OIDC exchange failure: {header}", file=sys.stderr)
+    sys.exit(1)
+
+
+def debug(msg: str):
+    print(f"::debug::{msg}", file=sys.stderr)
+
+
+def get_normalized_input(name: str) -> str | None:
+    name = f"INPUT_{name.upper()}"
+    if val := os.getenv(name):
+        return val
+    return os.getenv(name.replace("-", "_"))
+
+
+def assert_successful_audience_call(resp: requests.Response, domain: str):
+    if resp.ok:
+        return
+
+    match resp.status_code:
+        case 403:
+            # This index supports OIDC, but forbids the client from using
+            # it (either because it's disabled, ratelimited, etc.)
+            die(f"audience retrieval failed: repository at {domain} has OIDC disabled")
+        case 404:
+            # This index does not support OIDC.
+            die(
+                "audience retrieval failed: repository at "
+                f"{domain} does not indicate OIDC support",
+            )
+        case other:
+            # Unknown: the index may or may not support OIDC, but didn't respond with
+            # something we expect. This can happen if the index is broken, in maintenance mode,
+            # misconfigured, etc.
+            die(
+                "audience retrieval failed: repository at "
+                f"{domain} responded with unexpected {other}",
+            )
+
+
+repository_url = get_normalized_input("repository-url")
+is_pypi = not repository_url
+if is_pypi:
+    # Easy case: no explicit repository URL, which means we're using PyPI and we can just
```

Comment:
So, I realized something now... I clearly remember typing in a comment regarding this branching in general, but since it's nowhere to be found in the PR, I must've discarded it accidentally without actually posting :man_facepalming:

The gist of it was that we might not need this micro optimization at all.

The code above could be modified as follows:
```diff
- repository_url = get_normalized_input("repository-url")
+ repository_url = get_normalized_input("repository-url") or "https://pypi.org/legacy/"
```
Or, even better, the default could be set to `https://pypi.org/legacy/` in `action.yml`.

I think that one extra API call to ` /_/oidc/audience` would not DDoS PyPI because:
* publishing is not something that happens often in most projects
* said endpoint is _highly_ cacheable — the audience is the same for the entire instance deployment throughout its lifetime.

What's the benefit of dropping this if-branch? For me, it's better maintainability and more generic code path.

@woodruffw @di any objections?

Reply:
Agreed -- I think this endpoint will be cached, and even if not I don't think it'll have an appreciable impact on PyPI (or any other index) traffic.

I'll simplify here 🙂

Reply:
6e7606e